/* @flow */

/**
 * beforeAll
 */

beforeAll(() => {});
// $ExpectError
beforeAll();

/**
 * beforeEach
 */

beforeEach(() => {});
// $ExpectError
beforeEach();


/**
 * afterAll
 */

afterAll(() => {});
// $ExpectError
afterAll();

/**
 * afterEach
 */

afterEach(() => {});
// $ExpectError
afterEach();
